domain/src/main/java/life/qbic/apps/datamanager/services/UserRegistrationService.java

```diff
@@ -36,42 +51,151 @@ public UserRegistrationService(NotificationService notificationService, EventSto
-              @Override
-              public Class<UserRegistered> subscribedToEventType() {
-                return UserRegistered.class;
-              }
-
-              @Override
-              public void handleEvent(UserRegistered event) {
-                var notificationId = notificationService.newNotificationId();
-                var notification =
-                    Notification.create(
-                        UserRegistered.class.getSimpleName(),
-                        event.occurredOn(),
-                        notificationId,
-                        event);
-                eventStore.append(event);
-                notificationService.send(notification);
-              }
-            });
+    DomainEventPublisher.instance().subscribe(new DomainEventSubscriber<UserRegistered>() {
+      @Override
+      public Class<UserRegistered> subscribedToEventType() {
+        return UserRegistered.class;
+      }
+
+      @Override
+      public void handleEvent(UserRegistered event) {
+        eventStore.append(event);
+        sendNotification(event);
+      }
+
+      private void sendNotification(UserRegistered event) {
+        var notificationId = notificationService.newNotificationId();
+        var notification =
+            Notification.create(
+                event.getClass().getSimpleName(),
+                event.occurredOn(),
+                notificationId,
+                event);
+        notificationService.send(notification);
+      }
+    });
+
+    var userEmail = EmailAddress.from(email);
+    var userFullName = FullName.from(fullName);
+    var userPassword = EncryptedPassword.from(rawPassword);
+
+    if (userRepository.findByEmail(userEmail).isPresent()) {
+      return RegistrationResponse.failureResponse(new UserExistsException());
+    }
+
     // Trigger the user creation in the domain service
-    userDomainService.get().createUser(fullName, email, rawPassword);
+    userDomainService.get().createUser(userFullName, userEmail, userPassword);
+
     // Overwrite the password
     Arrays.fill(rawPassword, '-');
+    return RegistrationResponse.successResponse();
+  }
+
+  private RegistrationResponse validateInput(String fullName, String email, char[] rawPassword) {
+    List<RuntimeException> failures = new ArrayList<>();
+
+    try {
+      EmailAddress.from(email);
+    } catch (EmailValidationException e) {
+      failures.add(e);
+    }
+    try {
+      FullName.from(fullName);
+    } catch (InvalidFullNameException e) {
+      failures.add(e);
+    }
+    try {
+      EncryptedPassword.from(rawPassword);
+    } catch (PasswordValidationException e) {
+      failures.add(e);
+    }
+
+    if (failures.isEmpty()) {
+      return RegistrationResponse.successResponse();
+    }
+
+    if (failures.size() > 1) {
+      return RegistrationResponse.failureResponse(failures.toArray(RuntimeException[]::new));
+    } else {
+      return RegistrationResponse.failureResponse(failures.get(0));
+    }
```

Comment:
Can now be simplified
```suggestion
    return RegistrationResponse.failureResponse(failures.toArray(RuntimeException[]::new));
```